File: AWS_GUI_Dev/BAWGUI.Settings/ViewModels/SignalSignatureViewModel.vb
Imports BAWGUI.Core
Imports BAWGUI.Settings.ViewModels

Namespace ViewModel
    Public Class SignalSignatureViewModel
        Inherits ViewModelBase
        'Implements IDisposable
        Private _model As BAWGUI.Core.SignalSignatures
        Public Sub New()
            _model = New Core.SignalSignatures
            '_model.IsEnabled = True
            '_model.IsValid = True
            '_model.IsCustomSignal = False
            _model.PassedThroughDQFilter = 0
            _model.PassedThroughProcessor = 0
            '_model.IsNameTypeUnitChanged = False
            _model.SamplingRate = -1
            _model.Unit = "O"
        End Sub
        Public Sub New(name As String)
            _model = New Core.SignalSignatures
            _model.SignalName = name
            '_model.IsEnabled = True
            '_model.IsValid = True
            '_model.IsCustomSignal = False
            _model.PassedThroughDQFilter = 0
            _model.PassedThroughProcessor = 0
            '_model.IsNameTypeUnitChanged = False
            _model.SamplingRate = -1
            _model.Unit = "O"
        End Sub
        Public Sub New(name As String, pmu As String)
            _model = New Core.SignalSignatures
            _model.SignalName = name
            _model.PMUName = pmu
            '_model.IsEnabled = True
            '_model.IsValid = True
            '_model.IsCustomSignal = False
            _model.PassedThroughDQFilter = 0
            _model.PassedThroughProcessor = 0
            '_model.IsNameTypeUnitChanged = False
            _model.SamplingRate = -1
            _model.Unit = "O"
        End Sub
        Public Sub New(name As String, pmu As String, type As String)
            _model = New Core.SignalSignatures
            _model.SignalName = name
            _model.PMUName = pmu
            _model.TypeAbbreviation = type
            '_model.IsEnabled = True
            '_model.IsValid = True
            '_model.IsCustomSignal = False
            _model.PassedThroughDQFilter = 0
            _model.PassedThroughProcessor = 0
            '_model.IsNameTypeUnitChanged = False
            _model.SamplingRate = -1
            _model.Unit = "O"
        End Sub
        'Public Sub Dispose() Implements IDisposable.Dispose
        '    'Dispose(True)
        '    GC.SuppressFinalize(Me)
        'End Sub
        'Private managedResource As System.ComponentModel.Component
        'Private unmanagedResource As IntPtr
        'Protected disposed As Boolean = False
        'Private Sub Dispose(v As Boolean)
        '    If Not Me.disposed Then
        '        If v Then
        '            managedResource.Dispose()
        '        End If
        '        ' Add code here to release the unmanaged resource.
        '        unmanagedResource = IntPtr.Zero
        '        ' Note that this is not thread safe. 
        '    End If
        '    Me.disposed = True
        'End Sub
        Private _isValid As Boolean
        Public Property IsValid As Boolean
            Get
                Return _model.IsValid
            End Get
            Set(ByVal value As Boolean)
                _model.IsValid = value
                OnPropertyChanged()
            End Set
        End Property
        Private _PMUName As String
        Public Property PMUName As String
            Get
                Return _model.PMUName
            End Get
            Set(ByVal value As String)
                _model.PMUName = value
                OnPropertyChanged()
            End Set
        End Property
        Private _signalName As String
        Public Property SignalName As String
            Get
                Return _model.SignalName
            End Get
            Set(ByVal value As String)
                _model.SignalName = value
                If Not String.IsNullOrEmpty(value) AndAlso String.IsNullOrEmpty(OldSignalName) Then
                    OldSignalName = value
                End If
                OnPropertyChanged()
            End Set
        End Property
        Private _typeAbbreviation As String
        Public Property TypeAbbreviation As String
            Get
                Return _model.TypeAbbreviation
            End Get
            Set(ByVal value As String)
                _model.TypeAbbreviation = value
                If Not String.IsNullOrEmpty(value) AndAlso String.IsNullOrEmpty(OldTypeAbbreviation) Then
                    OldTypeAbbreviation = value
                End If
                OnPropertyChanged()
            End Set
        End Property
        Private _unit As String
        Public Property Unit As String
            Get
                Return _model.Unit
            End Get
            Set(ByVal value As String)
                _model.Unit = value
                If Not String.IsNullOrEmpty(value) AndAlso String.IsNullOrEmpty(OldUnit) Then
                    OldUnit = value
                End If
                OnPropertyChanged()
            End Set
        End Property
        Private _isChecked? As Boolean = False
        Public Property IsChecked As Boolean?
            Get
                Return _model.IsChecked
            End Get
            Set(ByVal value As Boolean?)
                _model.IsChecked = value
                OnPropertyChanged()
            End Set
        End Property
        Private _isEnabled As Boolean
        Public Property IsEnabled As Boolean
            Get
                Return _model.IsEnabled
            End Get
            Set(ByVal value As Boolean)
                _model.IsEnabled = value
                OnPropertyChanged()
            End Set
        End Property
        Private _isCustomSignal As Boolean
        Public Property IsCustomSignal As Boolean
            Get
                Return _model.IsCustomSignal
            End Get
            Set(ByVal value As Boolean)
                _model.IsCustomSignal = value
                OnPropertyChanged()
            End Set
        End Property

        Private _samplingRate As Integer
        Public Property SamplingRate As Integer
            Get
                Return _model.SamplingRate
            End Get
            Set(value As Integer)
                _model.SamplingRate = value
                OnPropertyChanged()
            End Set
        End Property

        Private _passedThroughDQFilter As Integer
        Public Property PassedThroughDQFilter As Integer
            Get
                Return _model.PassedThroughDQFilter
            End Get
            Set(value As Integer)
                _model.PassedThroughDQFilter = value
                OnPropertyChanged()
            End Set
        End Property

        Private _passedThroughProcessor As Integer
        Public Property PassedThroughProcessor As Integer
            Get
                Return _model.PassedThroughProcessor
            End Get
            Set(value As Integer)
                _model.PassedThroughProcessor = value
                OnPropertyChanged()
            End Set
        End Property

        Private _isNameTypeUnitChanged As Boolean
        Public Property IsNameTypeUnitChanged As Boolean
            Get
                Return _model.IsNameTypeUnitChanged
            End Get
            Set(ByVal value As Boolean)
                _model.IsNameTypeUnitChanged = value
                OnPropertyChanged()
            End Set
        End Property

        Public Overrides Function Equals(obj As Object) As Boolean
            If obj Is Nothing OrElse Not Me.GetType Is obj.GetType Then
                Return False
            End If
            Dim p As SignalSignatureViewModel = CType(obj, SignalSignatureViewModel)
            Return Me.PMUName = p.PMUName AndAlso Me.SignalName = p.SignalName AndAlso Me.TypeAbbreviation = p.TypeAbbreviation AndAlso Me.Unit = p.Unit AndAlso Me.OldSignalName = p.OldSignalName AndAlso Me.OldTypeAbbreviation = p.OldTypeAbbreviation AndAlso Me.OldUnit = p.OldUnit AndAlso Me.SamplingRate = p.SamplingRate
        End Function
        'Public Overrides Function GetHashCode() As Integer
        '    Return MyBase.GetHashCode()
        'End Function
        Public Shared Operator =(ByVal x As SignalSignatureViewModel, ByVal y As SignalSignatureViewModel) As Boolean
            Return x.PMUName = y.PMUName AndAlso x.SignalName = y.SignalName AndAlso x.TypeAbbreviation = y.TypeAbbreviation AndAlso x.Unit = y.Unit AndAlso x.OldSignalName = y.OldSignalName AndAlso x.OldTypeAbbreviation = y.OldTypeAbbreviation AndAlso x.OldUnit = y.OldUnit AndAlso x.SamplingRate = y.SamplingRate
        End Operator
        Public Shared Operator <>(ByVal x As SignalSignatureViewModel, ByVal y As SignalSignatureViewModel) As Boolean
            Return x.PMUName <> y.PMUName OrElse x.SignalName <> y.SignalName OrElse x.TypeAbbreviation <> y.TypeAbbreviation OrElse x.Unit <> y.Unit OrElse x.OldSignalName <> y.OldSignalName OrElse x.OldTypeAbbreviation <> y.OldTypeAbbreviation OrElse x.OldUnit <> y.OldUnit OrElse x.SamplingRate <> y.SamplingRate
        End Operator

        Public Function IsSignalInformationComplete() As Boolean
            Return Not String.IsNullOrEmpty(PMUName) AndAlso Not String.IsNullOrEmpty(SignalName) AndAlso Not String.IsNullOrEmpty(TypeAbbreviation) AndAlso Not String.IsNullOrEmpty(Unit) AndAlso (SamplingRate > 0)
        End Function

        Private _oldSignalName As String
        Public Property OldSignalName As String
            Get
                Return _model.OldSignalName
            End Get
            Set(ByVal value As String)
                _model.OldSignalName = value
                OnPropertyChanged()
            End Set
        End Property

        Private _oldUnit As String
        Public Property OldUnit As String
            Get
                Return _model.OldUnit
            End Get
            Set(ByVal value As String)
                _model.OldUnit = value
                OnPropertyChanged()
            End Set
        End Property

        Private _oldTypeAbbreviation As String
        Public Property OldTypeAbbreviation As String
            Get
                Return _model.OldTypeAbbreviation
            End Get
            Set(ByVal value As String)
                _model.OldTypeAbbreviation = value
                OnPropertyChanged()
            End Set
        End Property
    End Class

End Namespace

File: AWS_GUI_Dev/BAWGUI.Core/ViewModels/SignalSignatureViewModel.cs
using BAWGUI.Core.Models;
using BAWGUI.Core.Utilities;
using BAWGUI.Utilities;
using System.Collections.Generic;
using System;
using System.Collections.ObjectModel;
using System.Windows.Input;
using System.Linq;

namespace BAWGUI.Core
{
    public class SignalSignatureViewModel : ViewModelBase
    {
        // Implements IDisposable
        private SignalSignatures _model;
        public SignalSignatures Model
        {
            get { return _model; }
        }
        public SignalSignatureViewModel()
        {
            _model = new SignalSignatures();
            // _model.IsEnabled = True
            // _model.IsValid = True
            // _model.IsCustomSignal = False
            _model.PassedThroughDQFilter = 0;
            _model.PassedThroughProcessor = 0;
            // _model.IsNameTypeUnitChanged = False
            _model.SamplingRate = -1;
            _model.Unit = "O";
            DeleteASite = new RelayCommand(_deleteASite);
            AddASite = new RelayCommand(_addASite);
            _selectedLocation = Locations.FirstOrDefault();
            SelectedLocationIndex = 0;
        }
        public SignalSignatureViewModel(string name) : this()
        {
            _model.SignalName = name;
        }
        public SignalSignatureViewModel(string name, string pmu) : this(name)
        {
            _model.PMUName = pmu;
        }
        public SignalSignatureViewModel(string name, string pmu, string type) : this (name, pmu)
        {
            _model.TypeAbbreviation = type;
        }

        public bool? IsValid
        {
            get
            {
                return _model.IsValid;
            }
            set
            {
                _model.IsValid = value;
                OnPropertyChanged();
            }
        }
        public string PMUName
        {
            get
            {
                return _model.PMUName;
            }
            set
            {
                _model.PMUName = value;
                OnPropertyChanged();
            }
        }
        public string SignalName
        {
            get
            {
                return _model.SignalName;
            }
            set
            {
                _model.SignalName = value;
                if (!string.IsNullOrEmpty(value) && string.IsNullOrEmpty(OldSignalName))
                    OldSignalName = value;
                OnPropertyChanged();
            }
        }
        //private string _typeAbbreviation;
        public string TypeAbbreviation
        {
            get
            {
                return _model.TypeAbbreviation;
            }
            set
            {
                _model.TypeAbbreviation = value;
                if (!string.IsNullOrEmpty(value) && string.IsNullOrEmpty(OldTypeAbbreviation))
                    OldTypeAbbreviation = value;
                OnPropertyChanged();
            }
        }
        public string Unit
        {
            get
            {
                return _model.Unit;
            }
            set
            {
                _model.Unit = value;
                if (!string.IsNullOrEmpty(value) && string.IsNullOrEmpty(OldUnit))
                    OldUnit = value;
                OnPropertyChanged();
            }
        }
        public bool? IsChecked
        {
            get
            {
                return _model.IsChecked;
            }
            set
            {
                _model.IsChecked = value;
                OnPropertyChanged();
            }
        }
        public bool? IsEnabled
        {
            get
            {
                return _model.IsEnabled;
            }
            set
            {
                _model.IsEnabled = value;
                OnPropertyChanged();
            }
        }
        public bool? IsCustomSignal
        {
            get
            {
                return _model.IsCustomSignal;
            }
            set
            {
                _model.IsCustomSignal = value;
                OnPropertyChanged();
            }
        }

        //private int _samplingRate;
        public int SamplingRate
        {
            get
            {
                return _model.SamplingRate;
            }
            set
            {
                _model.SamplingRate = value;
                OnPropertyChanged();
            }
        }

        //private int _passedThroughDQFilter;
        public int PassedThroughDQFilter
        {
            get
            {
                return _model.PassedThroughDQFilter;
            }
            set
            {
                _model.PassedThroughDQFilter = value;
                OnPropertyChanged();
            }
        }

        //private int _passedThroughProcessor;
        public int PassedThroughProcessor
        {
            get
            {
                return _model.PassedThroughProcessor;
            }
            set
            {
                _model.PassedThroughProcessor = value;
                OnPropertyChanged();
            }
        }

        public bool? IsNameTypeUnitChanged
        {
            get
            {
                return _model.IsNameTypeUnitChanged;
            }
            set
            {
                _model.IsNameTypeUnitChanged = value;
                OnPropertyChanged();
            }
        }

        public override bool Equals(object obj)
        {
            if (obj == null || this.GetType() != obj.GetType())
                return false;
            SignalSignatureViewModel p = (SignalSignatureViewModel)obj;
            return this.PMUName == p.PMUName && this.SignalName == p.SignalName && this.TypeAbbreviation == p.TypeAbbreviation && this.Unit == p.Unit && this.OldSignalName == p.OldSignalName && this.OldTypeAbbreviation == p.OldTypeAbbreviation && this.OldUnit == p.OldUnit && this.SamplingRate == p.SamplingRate;
        }
        // Public Overrides Function GetHashCode() As Integer
        // Return MyBase.GetHashCode()
        // End Function
        public static bool operator ==(SignalSignatureViewModel x, SignalSignatureViewModel y)
        {
            if ((object)x == null)
            {
                return (object)y == null;
            }else if((object)y == null)
            {
                return false;
            }
            return x.PMUName == y.PMUName && x.SignalName == y.SignalName && x.TypeAbbreviation == y.TypeAbbreviation && x.Unit == y.Unit && x.OldSignalName == y.OldSignalName && x.OldTypeAbbreviation == y.OldTypeAbbreviation && x.OldUnit == y.OldUnit && x.SamplingRate == y.SamplingRate;
            //if (x is null && y is null)
            //{
            //    return true;
            //}
            //else if (x is null ^ y is null)
            //{
            //    return false;
            //}
            //else
            //{

            //    return x.PMUName == y.PMUName && x.SignalName == y.SignalName && x.TypeAbbreviation == y.TypeAbbreviation && x.Unit == y.Unit && x.OldSignalName == y.OldSignalName && x.OldTypeAbbreviation == y.OldTypeAbbreviation && x.OldUnit == y.OldUnit && x.SamplingRate == y.SamplingRate;
            //}
        }
        public static bool operator !=(SignalSignatureViewModel x, SignalSignatureViewModel y)
        {
            return !(x == y);
            //if ((object)x == null)
            //{
            //    return y != null;
            //}
            //return x.PMUName != y.PMUName || x.SignalName != y.SignalName || x.TypeAbbreviation != y.TypeAbbreviation || x.Unit != y.Unit || x.OldSignalName != y.OldSignalName || x.OldTypeAbbreviation != y.OldTypeAbbreviation || x.OldUnit != y.OldUnit || x.SamplingRate != y.SamplingRate;
            //if (x is null && y is null)
            //{
            //    return false;
            //}else if(x is null ^ y is null)
            //{
            //    return true;
            //}
            //else
            //{
            //    return x.PMUName != y.PMUName || x.SignalName != y.SignalName || x.TypeAbbreviation != y.TypeAbbreviation || x.Unit != y.Unit || x.OldSignalName != y.OldSignalName || x.OldTypeAbbreviation != y.OldTypeAbbreviation || x.OldUnit != y.OldUnit || x.SamplingRate != y.SamplingRate;
            //}
        }

        public bool IsSignalInformationComplete()
        {
            return !string.IsNullOrEmpty(PMUName) && !string.IsNullOrEmpty(SignalName) && !string.IsNullOrEmpty(TypeAbbreviation) && !string.IsNullOrEmpty(Unit) && (SamplingRate > 0);
        }

        //private string _oldSignalName;
        public string OldSignalName
        {
            get
            {
                return _model.OldSignalName;
            }
            set
            {
                _model.OldSignalName = value;
                OnPropertyChanged();
            }
        }

        //private string _oldUnit;
        public string OldUnit
        {
            get
            {
                return _model.OldUnit;
            }
            set
            {
                _model.OldUnit = value;
                OnPropertyChanged();
            }
        }

        //private string _oldTypeAbbreviation;
        public string OldTypeAbbreviation
        {
            get
            {
                return _model.OldTypeAbbreviation;
            }
            set
            {
                _model.OldTypeAbbreviation = value;
                OnPropertyChanged();
            }
        }
        //private List<double> _data;
        public List<double> Data
        {
            get { return _model.Data; }
            set
            {
                _model.Data = value;
                OnPropertyChanged();
            }
        }
        //private List<double> _timeStampNumber;
        public List<double> TimeStampNumber
        {
            get { return _model.TimeStampNumber; }
            set
            {
                _model.TimeStampNumber = value;
                OnPropertyChanged();
            }
        }
        //private List<double> _matlabTimeStampNumber;
        public List<double> MATLABTimeStampNumber
        {
            get { return _model.MATLABTimeStampNumber; }
            set
            {
                _model.MATLABTimeStampNumber = value;
                OnPropertyChanged();
            }
        }
        public SiteCoordinatesModel From
        {
            get { return _model.From; }
            set
            {
                _model.From = value;
                OnPropertyChanged();
            }
        }
        public SiteCoordinatesModel To
        {
            get { return _model.To; }
            set
            {
                _model.To = value;
                OnPropertyChanged();
            }
        }
        public SignalMapPlotType MapPlotType
        {
            get { return _model.MapPlotType; }
            set
            {
                if (_model.MapPlotType != value)
                {
                    _model.MapPlotType = value;
                    if (value == SignalMapPlotType.Line)
                    {
                        for (int index = Locations.Count; index < 2; index++)
                        {
                            Locations.Add(CoreUtilities.DummySiteCoordinatesModel);
                        }
                    }
                    if (value == SignalMapPlotType.Area)
                    {
                        for (int index = Locations.Count; index < 3; index++)
                        {
                            Locations.Add(CoreUtilities.DummySiteCoordinatesModel);
                        }
                    }
                    if (value == SignalMapPlotType.Dot)
                    {
                        if (Locations.Count == 0)
                        {
                            Locations.Add(CoreUtilities.DummySiteCoordinatesModel);
                        }
                        else
                        {
                            var keep = Locations[0];
                            Locations.Clear();
                            Locations.Add(keep);
                        }
                    }
                    SelectedLocation = Locations.LastOrDefault();
                    SelectedLocationIndex = Locations.Count - 1;
                    OnPropertyChanged();
                }
            }
        }
        public ObservableCollection<SiteCoordinatesModel> Locations
        {
            get { return _model.Locations; }
            set
            {
                _model.Locations = value;
                SelectedLocation = _model.Locations.FirstOrDefault();
                SelectedLocationIndex = 0;
                OnPropertyChanged();
            }
        }
        public ICommand DeleteASite { get; set; }
        private void _deleteASite(object obj)
        {
            var values = (object[])obj;
            var currentLocation = (SiteCoordinatesModel)values[0];
            SelectedLocationIndex = (int)values[1];
            _model.Locations.RemoveAt(SelectedLocationIndex);
            if (SelectedLocationIndex >= Locations.Count)
            {
                SelectedLocationIndex = Locations.Count - 1;
            }
            SelectedLocation = Locations[SelectedLocationIndex];
        }
        public ICommand AddASite { get; set; }
        private void _addASite(object obj)
        {
            _model.Locations.Add(CoreUtilities.DummySiteCoordinatesModel);
            SelectedLocation = Locations.LastOrDefault();
            SelectedLocationIndex = Locations.Count - 1;
        }
        private SiteCoordinatesModel _selectedLocation;
        public SiteCoordinatesModel SelectedLocation
        {
            get { return _selectedLocation; }
            set
            {
                _selectedLocation = value;
                OnPropertyChanged();
                OnLocationSelectionChanged(EventArgs.Empty);
            }
        }
        public int SelectedLocationIndex { get; set; }
        public event EventHandler LocationSelectionChanged;
        protected virtual void OnLocationSelectionChanged(EventArgs e)
        {
            LocationSelectionChanged?.Invoke(this, e);
        }
    }
}
